src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,380 @@
+};
+
+/*! \brief A functor that calculates the difference between the sample rate and the desired sample
+ * rows, given a cumulative gradient sum.
+ */
+class SampleRateDelta : public thrust::binary_function<float, size_t, float> {
+ public:
+  SampleRateDelta(common::Span<float> threshold, size_t n_rows, size_t sample_rows)
+      : threshold_(threshold), n_rows_(n_rows), sample_rows_(sample_rows) {}
+
+  XGBOOST_DEVICE float operator()(float gradient_sum, size_t row_index) const {
+    float lower = threshold_[row_index];
+    float upper = threshold_[row_index + 1];
+    float u = gradient_sum / static_cast<float>(sample_rows_ - n_rows_ + row_index + 1);
+    if (u > lower && u <= upper) {
+      threshold_[row_index + 1] = u;
+      return 0.0f;
+    } else {
+      return std::numeric_limits<float>::max();
+    }
+  }
+
+ private:
+  common::Span<float> threshold_;
+  size_t n_rows_;
+  size_t sample_rows_;
+};
+
+/*! \brief A functor that performs Poisson sampling, and scales gradient pairs by 1/p_i. */
+class PoissonSampling : public thrust::binary_function<GradientPair, size_t, GradientPair> {
+ public:
+  PoissonSampling(common::Span<float> threshold, size_t threshold_index, RandomWeight rnd)
+      : threshold_(threshold), threshold_index_(threshold_index), rnd_(rnd) {}
+
+  XGBOOST_DEVICE GradientPair operator()(const GradientPair& gpair, size_t i) {
+    // If the gradient and hessian are both empty, we should never select this row.
+    if (gpair.GetGrad() == 0 && gpair.GetHess() == 0) {
+      return gpair;
+    }
+    float combined_gradient = combine_(gpair);
+    float u = threshold_[threshold_index_];
+    float p = combined_gradient / u;
+    if (p >= 1) {
+      // Always select this row.
+      return gpair;
+    } else {
+      // Select this row randomly with probability proportional to the combined gradient.
+      // Scale gpair by 1/p.
+      if (rnd_(i) <= p) {
+        return gpair / p;
+      } else {
+        return GradientPair();
+      }
+    }
+  }
+
+ private:
+  common::Span<float> threshold_;
+  size_t threshold_index_;
+  RandomWeight rnd_;
+  CombineGradientPair combine_;
+};
+
+NoSampling::NoSampling(EllpackPageImpl* page) : page_(page) {}
+
+GradientBasedSample NoSampling::Sample(common::Span<GradientPair> gpair, DMatrix* dmat) {
+  return {dmat->Info().num_row_, page_, gpair};
+}
+
+ExternalMemoryNoSampling::ExternalMemoryNoSampling(EllpackPageImpl* page,
+                                                   size_t n_rows,
+                                                   const BatchParam& batch_param)
+    : batch_param_(batch_param),
+      page_(new EllpackPageImpl(batch_param.gpu_id, page->matrix.info, n_rows)) {}
+
+GradientBasedSample ExternalMemoryNoSampling::Sample(common::Span<GradientPair> gpair,
+                                                     DMatrix* dmat) {
+  if (!page_concatenated_) {
+    // Concatenate all the external memory ELLPACK pages into a single in-memory page.
```

Comment:
I think this option is here mostly for completeness' sake. If the whole dataset fits in GPU memory, then presumably you wouldn't want to use external memory; if it doesn't fit, you probably want to play around with sampling.

I have noticed that writing out external pages and then concatenating them together might allow you to train on slightly larger datasets versus keeping everything in memory, probably because of lower working memory requirement. Not sure how useful that is though.